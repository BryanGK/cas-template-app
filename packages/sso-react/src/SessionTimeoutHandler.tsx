import React, { useEffect, useState } from "react";
import LogoutWarningModal from "./LogoutWarningModal";

interface Props {
  modalDisplaySecondsBeforeLogout: number;

  extendSessionPath: string;
  sessionRemainingTimePath: string;
  logoutPath: string;

  // Callback for when the session has expired
  onSessionExpired: () => void;

  // Session-expired effect will recheck the session
  // if any of these values change.
  // e.g. with Next.js, use [router] where router = useRouter()
  resetOnChange: any[];
}

const SessionTimeoutHandler: React.FunctionComponent<Props> = ({
  modalDisplaySecondsBeforeLogout,
  extendSessionPath,
  sessionRemainingTimePath,
  logoutPath,
  onSessionExpired,
  resetOnChange,
}) => {
  const [showModal, setShowModal] = useState(false);

  // UNIX epoch (ms)
  const [sessionExpiresOn, setSessionExpiresOn] = useState(Infinity);

  const extendSession = async () => {
    const response = await fetch(extendSessionPath);
    if (response.ok) {
      const timeout = Number(await response.json());
      if (timeout > modalDisplaySecondsBeforeLogout) {
        setShowModal(false);
      }
      setSessionExpiresOn(timeout * 1000 + Date.now());
    }
  };

  useEffect(() => {
    let timeoutId: number;

    const checkSessionIdle = async () => {
      const response = await fetch(sessionRemainingTimePath);
      if (response.ok) {
        const timeout = Number(await response.json());

        setSessionExpiresOn(Date.now() + timeout * 1000);

        if (timeout > modalDisplaySecondsBeforeLogout) {
          setShowModal(false);
          timeoutId = setTimeout(() => {
            checkSessionIdle();
          }, (timeout - modalDisplaySecondsBeforeLogout) * 1000);
        } else if (timeout > 0) {
          // We display the modal and set a timeout to check again when the session is due to expire according to the server.
          // If the user has not extended their session by then we will redirect them (by invoking logoutOnSessionIdled() below)
          // If they do extend their session (or have in a different tab), the `checkSessionIdle()` call will branch into the first condition above, hide the modal,
          // and set another timeout to check the session idle when the modal is due to be displayed.

          setShowModal(true);
          timeoutId = setTimeout(() => {
            checkSessionIdle();
          }, timeout * 1000);
        } else {
          onSessionExpired();
        }
      } else {
        // if the response is not OK (i.e. 500)
        onSessionExpired();
      }
    };

    checkSessionIdle();

    // Return cleanup function
    return () => {
      clearTimeout(timeoutId);
    };
  }, resetOnChange);

  return (
    <>
      {showModal && (
        <LogoutWarningModal
          inactivityDelaySeconds={modalDisplaySecondsBeforeLogout}
          expiresOn={sessionExpiresOn}
          onExtendSession={extendSession}
          logoutPath={logoutPath}
        />
      )}
    </>
  );
};

export default SessionTimeoutHandler;
