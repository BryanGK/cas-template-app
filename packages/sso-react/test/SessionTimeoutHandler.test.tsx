import SessionTimeoutHandler from "../src/SessionTimeoutHandler";
import { mount } from "enzyme";
import React from "react";
import { act } from "react-dom/test-utils";

const existingFetch = global.fetch;

afterEach(() => {
  global.fetch = existingFetch;
  jest.useRealTimers();
  jest.clearAllMocks();
});

const setupFetchMock = (timeoutValue, response_override = {}) => {
  const fetchMock = jest.fn();
  fetchMock.mockImplementation(() => ({
    ok: true,
    json: async () => timeoutValue,
    ...response_override,
  }));

  global.fetch = fetchMock;
  return fetchMock;
};

describe("The Session Timeout Handler", () => {
  it("Shows the modal if there is less time left in the session than the delay", async () => {
    const secondsLeftInSession = 15;
    const displayDelayBeforeLogout = 30;

    jest.useFakeTimers();
    setupFetchMock(secondsLeftInSession);

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="extend"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={() => {}}
            resetOnChange={[{}]}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(componentUnderTest).toMatchSnapshot();
    expect(componentUnderTest.find("div.pg-modal-container")).toHaveLength(1);
  });

  it("renders the modal with the render function provided", async () => {
    const secondsLeftInSession = 15;
    const displayDelayBeforeLogout = 30;

    jest.useFakeTimers();
    setupFetchMock(secondsLeftInSession);

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="extend"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={() => {}}
            resetOnChange={[{}]}
            renderModal={({ remainingSeconds }) => (
              <div>Session will expire in {remainingSeconds} seconds</div>
            )}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(componentUnderTest).toMatchSnapshot();
  });

  it("Hides the modal if there is more time left in the session than the delay", async () => {
    const secondsLeftInSession = 45;
    const displayDelayBeforeLogout = 30;

    jest.useFakeTimers();
    setupFetchMock(secondsLeftInSession);

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="extend"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={() => {}}
            resetOnChange={[{}]}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(componentUnderTest.find("div.pg-modal-container").length).toBe(0);
  });

  it("Calls the onSessionExpired function if the session is expired", async () => {
    const mockExpiredCallback = jest.fn();

    const secondsLeftInSession = 0;
    const displayDelayBeforeLogout = 30;

    jest.useFakeTimers();
    setupFetchMock(secondsLeftInSession);

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="extend"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={mockExpiredCallback}
            resetOnChange={[{}]}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(mockExpiredCallback).toHaveBeenCalledOnce();
  });

  it("Calls the onSessionExpired function if the server replies with not ok", async () => {
    const mockExpiredCallback = jest.fn();

    const secondsLeftInSession = 0;
    const displayDelayBeforeLogout = 30;

    const fetchMock = setupFetchMock(secondsLeftInSession, { ok: false });

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="extend"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={mockExpiredCallback}
            resetOnChange={[{}]}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(fetchMock).toHaveBeenCalledWith("remaining");
    expect(mockExpiredCallback).toHaveBeenCalledOnce();
  });

  it("Calls the extendSessionPath endpoint when the user clicks the extend button", async () => {
    const secondsLeftInSession = 15;
    const displayDelayBeforeLogout = 30;

    jest.useFakeTimers();
    setupFetchMock(secondsLeftInSession);

    let componentUnderTest;
    await act(async () => {
      componentUnderTest = mount(
        <div>
          <SessionTimeoutHandler
            modalDisplaySecondsBeforeLogout={displayDelayBeforeLogout}
            extendSessionPath="my-test-extend-session-URL"
            logoutPath="logout"
            sessionRemainingTimePath="remaining"
            onSessionExpired={() => {}}
            resetOnChange={[{}]}
          />
        </div>
      );
    });

    await componentUnderTest.update();

    expect(componentUnderTest.find("div.pg-modal-container").length).toBe(1);

    const fetchMock = setupFetchMock(999);

    const clickRefeshHandler = componentUnderTest
      .find("button#logout-warning-modal-remain-active-button")
      .prop("onClick");

    await act(async () => {
      await clickRefeshHandler();
    });

    await componentUnderTest.update();

    expect(componentUnderTest.find("div.pg-modal-container").length).toBe(0);
    expect(fetchMock).toHaveBeenCalledWith("my-test-extend-session-URL");
  });
});
