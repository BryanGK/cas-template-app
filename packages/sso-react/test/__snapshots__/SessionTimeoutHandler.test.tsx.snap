// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Session Timeout Handler Shows the modal if there is less time left in the session than the delay 1`] = `
<div>
  <div
    className="sc-iCfMLu sc-furwcr kQXSYc gDlUxH pg-modal-container"
    id="logout-warning-modal"
    size="lg"
  >
    <div
      className="sc-pVTFL fdkSXM pg-modal-main"
    >
      <header
        className="sc-jrQzAO eeqVZV pg-modal-header h4"
      >
        Inactivity Logout Warning
      </header>
      <div
        className="sc-kDTinF cIwtan pg-modal-content"
        style={
          Object {
            "padding": "2em",
          }
        }
      >
        <div
          className="sc-fFeiMQ bgyiGU pg-grid-container"
          cols={12}
        >
          <div
            className="sc-bkkeKt sc-ieecCq hOEAKj fEyLc pg-grid-row"
          >
            Your session is about to expire due to inactivity over
             
            0
             minutes.
          </div>
          <div
            className="sc-bkkeKt sc-ieecCq hOEAKj fEyLc pg-grid-row"
          >
            You will be logged out in 
            15
             seconds.
          </div>
        </div>
      </div>
      <footer
        className="sc-iqseJM cULDjt pg-modal-footer"
      >
        <form
          action="logout"
          method="post"
        >
          <button
            className="sc-kfPuZi jgizTO pg-button"
            id="logout-warning-modal-logout-button"
            type="submit"
          >
            Logout
          </button>
        </form>
        <button
          className="sc-kfPuZi epUors pg-button"
          id="logout-warning-modal-remain-active-button"
          onClick={[Function]}
        >
          Remain active
        </button>
      </footer>
    </div>
  </div>
</div>
`;

exports[`The Session Timeout Handler renders the modal with the render function provided 1`] = `
<div>
  <div>
    Session will expire in 
    15
     seconds
  </div>
</div>
`;
